ui/src/criteria/attribute.tsx

```diff
@@ -81,73 +136,216 @@ function AttributeEdit(props: AttributeEditProps) {
+
+    // two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number | string>(
+      props.data.min || minBound
+    );
+    const [maxInputValue, setMaxInputValue] = useState<number | string>(
+      props.data.max || maxBound
+    );
+
+    const updateValue = (newMin: number, newMax: number) => {
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleMinInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin = event.target.value === "" ? "" : Number(event.target.value);
+      setMinInputValue(newMin); // Make sure empty input won't get changed.
+      newMin = typeof newMin === "number" ? newMin : 0;
+      const maxValue = props.data?.max || maxBound;
 
-  const selectionIndex = (hint: tanagra.EnumHintValue) =>
-    props.data.selected.findIndex(
-      (row) => row.id === hint.attributeValue?.int64Val
+      newMin = Math.max(minBound, newMin);
+      newMin = Math.min(maxValue, newMin);
+      updateValue(newMin, maxValue);
+    };
+    const handleMinInputBlur = () => {
+      let newMin = typeof minInputValue === "number" ? minInputValue : 0;
+      const maxValue = props.data?.max || maxBound;
+      newMin = Math.max(newMin, minBound);
+      newMin = Math.min(newMin, maxValue);
+      setMinInputValue(newMin);
+    };
+
+    const handleMaxInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMax = event.target.value === "" ? "" : Number(event.target.value);
+      setMaxInputValue(newMax); // Make sure empty input won't get changed.
+      newMax = typeof newMax === "number" ? newMax : 0;
+      const minValue = props.data?.min || minBound;
+
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      updateValue(minValue, newMax);
+    };
+    const handleMaxInputBlur = () => {
+      let newMax = typeof maxInputValue === "number" ? maxInputValue : 0;
+      const minValue = props.data?.min || minBound;
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      setMaxInputValue(newMax);
+    };
+
+    return (
+      <Box sx={{ width: "30%", minWidth: 300, margin: 2 }}>
+        <Grid container spacing={2}>
+          <Grid item>
+            <Input
+              value={minInputValue}
+              size="medium"
+              onChange={handleMinInputChange}
+              onBlur={handleMinInputBlur}
+              inputProps={{
+                step: ((maxBound - minBound) / 20) | 0,
+                min: minBound,
```

Comment:
Do these actually work for an input? What happens if you backspace and the value goes below min? Seems like it either interferes with editing or it can't actually work so it should be removed unless I'm misunderstanding the behavior.

Reply:
![Screen Shot 2022-04-26 at 4 53 52 PM](https://user-images.githubusercontent.com/100877272/165390600-df39e8f5-f079-4c86-9b33-4e996feea504.png)
this is I backspace, and the mouse focus on the input number.
![Screen Shot 2022-04-26 at 4 52 27 PM](https://user-images.githubusercontent.com/100877272/165390320-7ae41ee2-8af7-4145-a946-8c3eccd36b9c.png)
after the mouse click other place and it does not focus on the input blank. it will automatically go to the number which we support(in this case, it is the min number 1930)

Reply:
Hi Tim, I just double checked. The `min` and `max` props here actually can bound the range of input box during pressing the up or down button at the side (or simply up or down key on keyboard). Like the following screenshot. Without these two props, the input box can either be increased or decreased without bound until the user click somewhere else, which trigger the `onBlur` callback. I think keeping these two bounds here will be better.
<img width="524" alt="Screen Shot 2022-04-28 at 12 18 46 AM" src="https://user-images.githubusercontent.com/100877272/165675735-88f3215c-e7af-49c8-8191-9ca691aa2a31.png">

Reply:
Okay, good to leave them in then.